Fix for issue #39827

taleks opened the pull request:
*Cause of the issue*
    
While preparing for `trans_intrinsic_call()` invoke arguments are processed with `trans_argument()` method which excludes zero-sized types from argument list (to be more correct - all arguments for which `ArgKind` is `Ignore` are filtered out). As result `volatile_store()` intrinsic gets one argument instead of expected address and value.
    
*How it is fixed*
 
Modification of the `trans_argument()` method may cause side effects, therefore change was implemented in `volatile_store()` intrinsic building code itself. Now it checks function signature and if it was specialised with zero-sized type, then emits `C_nil()` instead of accessing non-existing second argument.

rust-highfive commented:
Thanks for the pull request, and welcome! The Rust team is excited to review your changes, and you should hear from @pnkfelix (or someone else) soon.

If any changes to this PR are deemed necessary, please add them as extra commits. This ensures that the reviewer can see what has changed since they last reviewed the code. Due to the way GitHub handles out-of-date commits, this should also make it reasonably obvious what issues have or haven't been addressed. Large or tricky changes may require several passes of review and changes.

Please see [the contribution instructions](https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md) for more information.

arielb1 commented:
Nice PR!

I'm not sure we want the "suspicious monomorphization" warning - it can spam the user in dead generic code. r=me without it.

arielb1 commented on src/libcore/ptr.rs:
```diff
@@ -384,6 +384,12 @@ pub unsafe fn write_unaligned<T>(dst: *mut T, src: T) {
 /// over time. That being said, the semantics will almost always end up pretty
 /// similar to [C11's definition of volatile][c11].
 ///
+/// Compiler shouldn't change relative order or number of volatile memory
```
eng:

The compiler shouldn't change the relative order or number of volatile memory operations. However, volatile memory operations on zero-sized types (e.g. if a zero-sized type is passed to `read_volatile`) are no-ops and may be ignored.

taleks commented:
Thanks for comments.
I've removed warnings related code from `librustc_trans/intrinsic.rs` and updated changes in documentation  for `write_volatile()`/`read_volatile()`.

arielb1 commented:
@bors r+

bors commented:
:pushpin: Commit faf6b84 has been approved by `arielb1`

<!-- @bors r=arielb1 faf6b84304dbc96f7bf142973f394f820e390a4c -->

bors commented:
:hourglass: Testing commit faf6b84304dbc96f7bf142973f394f820e390a4c with merge f3cf2062015200c0705ec2820dd02a7b9dc9bb22...

bors commented:
:sunny: Test successful - [status-appveyor](https://ci.appveyor.com/project/rust-lang/rust/build/1.0.4276), [status-travis](https://travis-ci.org/rust-lang/rust/builds/264192470?utm_source=github_status&utm_medium=notification)
Approved by: arielb1
Pushing f3cf2062015200c0705ec2820dd02a7b9dc9bb22 to master...
